src/net/thread.rs

```diff
@@ -54,11 +81,19 @@ impl Default for Thread {
     }
 }
 
-impl Drop for Thread {
+impl Drop for Pool {
     fn drop(&mut self) {
-        info!("Terminating thread {:?}", self.worker);
-        if let Ok(()) = self.sender.send(Message::Terminate) {
-            if let Some(thread) = self.worker.thread.take() {
+        for _ in &self.workers {
+            if self.sender.send(Message::Terminate).is_err() {
+                warn!("Receiver channel is closed.");
+                // If SendError occurs, that means the receiver is deallocated and all following send() calls will fail.
+                break;
+            }
+        }
+
+        for worker in &mut self.workers {
+            info!("Terminating thread {:?}", worker);
+            if let Some(thread) = worker.thread.take() {
                 thread.join().unwrap();
             }
         }
```

Comment:
Thinking more about this, it might be worth breaking this down a little more.

What you are looking for here is a cleanup function that ensures that you don't get zombie threads.  For that, sending the Terminate instruction is probably sufficient.  The reason that you want to call join() is to ensure that you get a clean shutdown.  Just sending N terminate instructions won't ensure that each of the threads has time to finish open tasks, so a join is necessary there.

So my suggestion would be to have the Drop implementation only send the Terminate instruction.  Then, if the Pool controller is dropped (by accident, I hope), the threads are not left dangling.  But if you want a graceful shutdown, then you need the join.

So I think you have two functions: terminate() and shutdown().  The first sends the terminate message to all the workers without waiting for them to come back.  The second calls terminate() and then waits for all of the threads to end.  drop() only really needs to do the first.  Then you don't have a call that potentially blocks for a long time in a drop implementation.